sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1304,13 +1306,29 @@ private[lf] object SBuiltin {
   // matches the template type, and then return the SAny internal value.
   final case class SBFromInterface(
       tplId: TypeConName
-  ) extends SBuiltinPure(1) {
-    override private[speedy] def executePure(args: util.ArrayList[SValue]): SOptional = {
+  ) extends SBuiltin(1) {
+    override private[speedy] def execute[Q](
+        args: util.ArrayList[SValue],
+        machine: Machine[Q],
+    ): Control[Q] = {
       val (tyCon, record) = getSAnyContract(args, 0)
+
       if (tplId == tyCon) {
-        SOptional(Some(record))
+        Control.Value(SOptional(Some(record)))
+      } else if (tplId.qualifiedName == tyCon.qualifiedName) {
+        val tplIdTemplateVersion = machine.tmplId2TxVersion(tplId)
+        val oTplIdPkgName = machine.tmplId2PackageName(tplId, tplIdTemplateVersion)
+        val tyConTemplateVersion = machine.tmplId2TxVersion(tyCon)
+        val oTyConPkgName = machine.tmplId2PackageName(tyCon, tyConTemplateVersion)
+        (oTplIdPkgName, oTyConPkgName) match {
+          case (Some(tplIdPkgName), Some(tyConPkgName)) if tplIdPkgName == tyConPkgName =>
+            importValue(machine, tplId, record.toUnnormalizedValue) { templateArg =>
+              Control.Value(SOptional(Some(templateArg)))
+            }
+          case _ => Control.Value(SOptional(None))
+        }
       } else {
-        SOptional(None)
+        Control.Value(SOptional(None))
```

Comment:
Not `Control.Error`?

Reply:
No, explicitly not that - fromInterface should give back None if the given template is not the "same" as what it holds, where "same" here means either same template id (+package id), or same qualified name + same package name + both support upgrades.
If there is an upgrade failure, the importValue will error.